### DISCUSSION TITLE
How to use NativeModules in react-native-tvos?

### DISCUSSION THREAD

**Discussion by avishka-codimite** [react-native-tvos/react-native-tvos#401] [Q&A] [answered]
I had to face a problem when using NativeModules in tvos library. Seems to be NativeModules not supported in tvos library.  

> TypeError: null is not an object (evaluating '_reactNative.NativeModules.RNHello.addEvent')

**Comment by douglowder** [comment, marked as answer]
This means that `RNHello` or `RNHello.addEvent` is not actually implemented in native code.

**Comment by freeboub** [comment]
Or RNhello is not well imported!

**Comment by avishka-codimite** [comment]
Hi, yeah. there was a mistake. thank you for pointing out that.  NativeModules is working fine. I had implemented a native code for preventing the screen saver but I can't figure out the reason for the following error. 

>  TypeError: null is not an object (evaluating '_reactNative.NativeModules.RNKeepAwake.activate')

source: [@sayem314/react-native-keep-awake](https://github.com/sayem314/react-native-keep-awake)

**RNKeepAwake.h file**

```
#import <React/RCTBridgeModule.h>

@interface RNKeepAwake : NSObject <RCTBridgeModule>
@end
```

**RNKeepAwake.m file**

```
#import "RNKeepAwake.h"
#import <UIKit/UIKit.h>


@implementation RNKeepAwake

RCT_EXPORT_MODULE();

RCT_EXPORT_METHOD(activate)
{
  dispatch_async(dispatch_get_main_queue(), ^{
    [[UIApplication sharedApplication] setIdleTimerDisabled:YES];
  });
}

RCT_EXPORT_METHOD(deactivate)
{
  dispatch_async(dispatch_get_main_queue(), ^{
    [[UIApplication sharedApplication] setIdleTimerDisabled:NO];
  });
}

@end
```

**App.js file**

```
import ReactNative, { NativeModules} from 'react-native';

const App: () => Node = () => {
const [openArtwork, setOpenArtwork] = useState(false);

  const openHandler = () => {
    setOpenArtwork(true);
    NativeModules.RNKeepAwake.activate();
  };

  const closeHandler = () => {
    setOpenArtwork(false);
    NativeModules.RNKeepAwake.deactivate();
  };
  return (
    <SafeAreaView style={styles.background}>
      {!openArtwork && (
        <TouchableHighlight onPress={openHandler}>
          <View style={styles.btn}>
            <Text style={styles.btnText}>Press</Text>
          </View>
        </TouchableHighlight>
      )}

      {openArtwork && (
        <TouchableHighlight onPress={closeHandler}>
          <Image
            source={{
              uri: 'https://images.pexels.com/photos/911758/pexels-photo-911758.jpeg?auto=compress&cs=tinysrgb&w=800',
            }}
            style={styles.img}
          />
        </TouchableHighlight>
      )}
    </SafeAreaView>
  );
};
```

**Reply by douglowder** [reply]
Sorry for the delay in response.

I see that the module you are using (`react-native-keep-awake`) does not support tvOS by default. That may be the problem. I'll have a look at creating a simple example to see if I can reproduce this issue.

**Reply by avishka-codimite** [reply]
thanks @douglowder

**Reply by douglowder** [reply]
This works: https://github.com/douglowder/TVKeepAwake 

- Has the required patch to `react-native-keep-awake`
- Includes the `<KeepAwake />` component in `App.js` that invokes the native API on mount

**Reply by avishka-codimite** [reply]
I tested it on Apple tv. it's working. thank you @douglowder
